File: PrefixSum/2615. Sum of Distances/solution.py
# Brute Force, TLE
class Solution_TLE:
    def distance(self, nums: List[int]) -> List[int]:
        n = len(nums)

        group = defaultdict(list)
        for i, num in enumerate(nums):
            group[num].append(i)
            
        dist = defaultdict(lambda: [0] * n)
        for k, vals in group.items():
            m = len(vals)
            for i in range(m):
                SUM = 0
                for j in range(m):
                    SUM += abs(vals[i]-vals[j])
                dist[k][vals[i]] = SUM
        
        arr = [0] * n
        for i in range(n):
            if not group[nums[i]]: continue
            arr[i] = dist[nums[i]][i]
        return arr

class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n

        # left to right to calculate left_part
        count = defaultdict(int)
        prefix_sum = defaultdict(int)
        for i, num in enumerate(nums):
            res[i] += count[num] * i - prefix_sum[num]
            prefix_sum[num] += i
            count[num] += 1

        # right to left to calculate right_part
        count = defaultdict(int)
        post_sum = defaultdict(int)
        for i in range(n-1, -1, -1):
            res[i] += post_sum[nums[i]] - count[nums[i]] * i
            post_sum[nums[i]] += i
            count[nums[i]] += 1
        return res

File: PrefixSum/2615. Sum of Distances/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

using LL = long long;

class Solution {
public:
    vector<long long> distance(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, vector<LL>> indices;

        for (int i = 0; i < n; i++) {
            indices[nums[i]].push_back(i);
        }

        vector<LL> res(n, 0);
        
        for (auto& [val, idxs] : indices) {
            int m = idxs.size();
            if (m == 1) continue;
            
            // Create prefix sum using partial_sum
            vector<LL> presum(m);
            partial_sum(idxs.begin(), idxs.end(), presum.begin());
            
            for (int i = 0; i < m; i++) {
                LL idx = idxs[i];
                
                // Left: sum of distances to indices before i
                LL leftSum = i * idx - (i > 0 ? presum[i - 1] : 0);
                
                // Right: sum of distances to indices after i
                LL rightSum = (presum[m - 1] - presum[i]) - (m - 1 - i) * idx;
                
                res[idx] = leftSum + rightSum;
            }
        }

        return res;
    }
};